### PR TITLE
Add support for downloading zip and tar archives of repositories.

### PR DESCRIPTION
**Pull request by chids**
👋, found myself needing to download zip/tar archives of repository content which wasn't supported so mucked about with getting that in place.

Now, there's a few things to note here which I'm not sure if/how to tackle:

1. The wiremock setup for the unit tests seems to use a HTTP client that _doesn't_ follow redirects whereas the one used by default seems to....bug? 🐛
2. The archive [endpoints](https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#download-a-repository-archive-zip) return a `302` redirect to the actual archive which is hosted at `https://codeload.github.com/<owner>/<repo>/legacy.<tar.gz|zip>/<ref>` but the redirect doesn't seem to be captured properly by wiremock and the download itself isn't captured at all even when adding manual logic to follow the redirect
3. Then, more of an FYI, I decided to "drop down" to `Requester.client` in order to be able to stream the content directly into a `StreamConsumer` since `Requester.fetchStream()` reads everything into memory which didn't seem ideal for dealing with entire repository archives

### CODE REVIEW

**Comment by bitwiseman** [comment]
@chids 

Item 2 will require an update to the test framework to fix, but it should be pretty easy - adding another mock server to the list and then recording.  You can do that as part of this PR if you like.

**Review by bitwiseman — CHANGES_REQUESTED** [review]
@timja @chids 

I updated this with new library-wide functional interfaces.  Also fixed the `codeload` and redirect issue. 

The main question I have is whether it is better to have the public `readZip` and `readTar` methods take a `Consumer` or a `Function`.  This matters because implementing one now and adding the other later may result in compilation errors - the compiler may not be able to distinguish between the two types for some lambda inputs.

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 1]
```diff
@@ -2865,6 +2866,56 @@ public GHTagObject createTag(String tag, String message, String object, String t
...
+     * A functional interface, equivalent to {@link java.util.function.Consumer} but that allows throwing
+     * {@link IOException}
+     */
+    @FunctionalInterface
+    public interface StreamConsumer {
+        void accept(InputStream stream) throws IOException;
+    }
+
+    private void downloadArchive(String type, Optional<String> ref, StreamConsumer sink) throws IOException {
+        requireNonNull(sink, "Sink must not be null");
```
[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/org/kohsuke/github/GHRepository.java:2997]: the annotation doesn't actually enforce it at run time though as far as I know?
they provide hints for IDE and static analysis tools

**Review by timja — APPROVED** [review]
not sure either way on whether consumer of function is better

**Comment by bitwiseman** [comment]
@timja  Thanks for taking the time to review.

**Review by bitwiseman — APPROVED** [review]

**Pull request merged**